### PR TITLE
add index-based softmax cross entropy loss

### PR DESCRIPTION
**Pull request by marcrasi**
As per discussion in https://groups.google.com/a/tensorflow.org/d/msg/swift/ASQJJkjf-xU/7bl6HOvdBQAJ

### CODE REVIEW

**Comment by marcrasi** [comment]
Oops, the implementation that I copied was not generic and I did not notice that. Will update it to be generic.

**Comment by marcrasi** [comment]
Updated to use generic Scalar.
